libp2p/connmanager.nim

```diff
@@ -459,25 +467,28 @@ proc trackOutgoingConn*(c: ConnManager,
   ## exception
   ##
 
-  trace "Tracking outgoing connection", count = c.outSema.count,
-                                        max = c.outSema.size
+  trace "Tracking outgoing connection", count = c.connSema.count,
+                                        max = c.connSema.size
 
-  if not c.outSema.tryAcquire():
-    trace "Too many outgoing connections!", count = c.outSema.count,
-                                            max = c.outSema.size
+  if c.maxConnections > 0 and c.conns.len >= c.maxConnections:
+    trace "Too many outgoing connections!", count = c.connSema.count,
+                                            max = c.connSema.size
     raise newTooManyConnectionsError()
 
+  # always count against the semaphore regardless
+  # if the semaphore limit has been reached
+  asyncSpawn c.connSema.acquire()
```

Comment:
Not a big fan of this approach, if the acquire blocks, we may release before acquiring
In the current setup that would worst case scenario allow one peer in, but still not great

A semaphore with a forceAcquire would be better, I think

Basically, forceAcquire->release on a full semaphore needs to be a noop, instead of popping one